sample/sample_test.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		t.Fatal(err)
+	}
+	expectedlogits := []float64{math.NaN(), math.NaN(), math.NaN(), math.NaN(), 1, 2, 4}
+	if !floats.Same(logits, expectedlogits) {
+		t.Fatalf("logits: %v, expectedlogits: %v", logits, expectedlogits)
+	}
+	logits, err = TopK(0).Sample([]float64{-3, -2, -1, 0, 1, 2, 4})
+	if err == nil {
+		t.Fatalf("expected error for k=0, got %v", logits)
+	}
+}
+
+func TestTopP(t *testing.T) {
+	logits, err := TopP(0.9).Sample([]float64{-3, -2, -1, 0, 1, 2, 4})
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := []float64{math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 2, 4}
+	if !floats.Same(logits, want) {
+		t.Fatalf("got: %v, want: %v", logits, want)
+	}
+	logits, err = TopP(1.0).Sample([]float64{-3, -2, -1, 0, 1, 2, 4})
+	if err == nil {
+		t.Fatalf("expected error for p=1.0, got %v", logits)
+	}
+	logits, err = TopP(0.0).Sample([]float64{-3, -2, -1, 0, 1, 2, 4})
+	if err == nil {
+		t.Fatalf("expected error for p=0.0, got %v", logits)
+	}
+}
+
+func TestMinP(t *testing.T) {
+	logits, err := MinP(0.2).Sample([]float64{-3, -2, -1, 0, 1, 2, 3, 4})
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := []float64{math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), math.NaN(), 3, 4}
+	if !floats.Same(logits, want) {
+		t.Fatalf("got: %v, want: %v", logits, want)
+	}
+	logits, err = MinP(1.0).Sample([]float64{-3, -2, -1, 0, 1, 2, 3, 4})
+	if err == nil {
+		t.Fatalf("expected error for p=1.0, got %v", logits)
+	}
+	logits, err = MinP(0.0).Sample([]float64{-3, -2, -1, 0, 1, 2, 3, 4})
+	if err == nil {
+		t.Fatalf("expected error for p=0.0, got %v", logits)
+	}
+}
+
+func TestWeighed(t *testing.T) {
+	logits, err := Weighed().Sample([]float64{math.NaN(), 2, math.NaN(), math.NaN()})
+	if err != nil {
+		t.Fatal(err)
+	}
+	want := []float64{1}
+	if !floats.Equal(logits, want) {
+		t.Fatalf("got: %v, want: %v", logits, want)
+	}
+	logits, err = Weighed().Sample([]float64{math.NaN(), math.NaN(), math.NaN()})
+	if err == nil {
+		t.Fatalf("expected error for no valid tokens, got %v", logits)
+	}
+}
+
+func TestSample(t *testing.T) {
+	input := []float64{1, 2, 3, 4}
+	want := []float64{1, 2, 3, 4}
+
+	var callOrder []int
+	mock1 := &mockSampler{
+		id:         1,
+		callOrder:  &callOrder,
+		returnVals: want,
+	}
+	mock2 := &mockSampler{
+		id:         2,
+		callOrder:  &callOrder,
+		returnVals: want,
+	}
+	mock3 := &mockSampler{
+		id:         3,
+		callOrder:  &callOrder,
+		returnVals: want,
+	}
+
+	got, err := Sample(input, mock1, mock2, mock3)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	if !slices.Equal(callOrder, []int{1, 2, 3}) {
+		t.Errorf("got %v, want %v", callOrder, []int{1, 2, 3})
+	}
+
+	if !floats.Equal(got, want) {
+		t.Errorf("got %v, want %v", got, want)
+	}
+
+	errMock := &mockSampler{
+		returnErr: fmt.Errorf("mock error"),
+	}
+	_, err = Sample(input, mock1, errMock, mock2)
+	if err == nil {
+		t.Error("Expected error from sampler")
+	}
+}
+
+type mockSampler struct {
```

Comment:
```suggestion
type testSampler struct {
```

might be a better name